File: mobile_frontend/bumpbox_panel/lib/services/item_api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/item.dart';
import '../config/pricing_config.dart';

/// Service for fetching items from the backend API
class ItemApiService {
  /// Fetch the latest listed item from the backend
  ///
  /// Returns null if no item is found or an error occurs.
  /// The backend returns the most recently listed item.
  static Future<Item?> fetchLatestItem() async {
    try {
      final response = await http.get(
        Uri.parse(ApiConfig.getItemUrl),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 404) {
        // No item found
        print('[ItemApiService] No item found in backend');
        return null;
      }

      if (response.statusCode != 200) {
        throw Exception(
          'Failed to fetch item: ${response.statusCode} ${response.body}',
        );
      }

      final jsonData = jsonDecode(response.body);

      // Backend response structure:
      // {
      //   "status": true/false,  // true if sold, false if not sold
      //   "message": "...",
      //   "data": { item fields }
      // }

      if (jsonData['data'] == null) {
        print('[ItemApiService] No item data in response');
        return null;
      }

      // Extract the status field to determine if item is sold
      final isSold = jsonData['status'] == true;

      return _parseItemFromBackend(jsonData['data'], isSold: isSold);
    } catch (e) {
      print('[ItemApiService] Error fetching item: $e');
      return null;
    }
  }

  /// Parse backend item data into Item model
  ///
  /// Backend provides:
  /// - itemid: database ID
  /// - item_name: item name
  /// - price: current/starting price
  /// - description: item description (if available)
  /// - datetime_expire: expiration timestamp
  /// - sale_status: 0 (not sold) or 1 (sold)
  /// - paymentLink: Stripe payment link URL
  ///
  /// We map these to the Item model's fields:
  /// - id: from itemid
  /// - name: from item_name
  /// - description: from description or default
  /// - startingPrice: from price
  /// - floorPrice: calculated as 50% of starting price (backend doesn't store this yet)
  /// - listedAt: calculated from datetime_expire - listingDuration
  /// - listingDuration: from pricing config
  /// - paymentLink: from paymentLink
  /// - isSold: whether the item has been sold (from status field)
  static Item _parseItemFromBackend(
    Map<String, dynamic> data, {
    bool isSold = false,
  }) {
    final itemId = data['itemid']?.toString() ?? 'unknown';
    final itemName = data['item_name']?.toString() ?? 'Unknown Item';
    final price = _parsePrice(data['price']);
    final description =
        data['description']?.toString() ??
        'High-quality item available for purchase';
    final paymentLink = data['paymentLink']?.toString();

    // Parse expiration date to calculate listing time
    final expirationDate = _parseDateTime(data['datetime_expire']);

    // Calculate listedAt as (expirationDate - listingDuration)
    final listedAt = expirationDate.subtract(PricingConfig.listingDuration);

    return Item(
      id: itemId,
      name: itemName,
      description: description,
      startingPrice: price * 1.15, // Add 15% markup for starting price
      floorPrice: price,
      listedAt: listedAt,
      listingDuration: PricingConfig.listingDuration,
      paymentLink: paymentLink,
      isSold: isSold,
    );
  }

  /// Parse price from backend (can be string or number)
  static double _parsePrice(dynamic priceData) {
    if (priceData == null) return 100.0; // Default fallback

    if (priceData is num) {
      return priceData.toDouble();
    }

    if (priceData is String) {
      return double.tryParse(priceData) ?? 100.0;
    }

    return 100.0;
  }

  /// Parse datetime from backend
  ///
  /// Backend stores datetime_expire as MySQL DATETIME format
  /// (e.g., "2024-01-15 14:30:00")
  static DateTime _parseDateTime(dynamic dateData) {
    if (dateData == null) {
      // Default to 7 days from now if no expiration
      return DateTime.now().add(PricingConfig.listingDuration);
    }

    if (dateData is String) {
      try {
        // Try parsing ISO 8601 format
        return DateTime.parse(dateData);
      } catch (e) {
        // Try parsing MySQL datetime format (YYYY-MM-DD HH:MM:SS)
        try {
          final parts = dateData.split(' ');
          if (parts.length == 2) {
            final dateParts = parts[0].split('-');
            final timeParts = parts[1].split(':');

            if (dateParts.length == 3 && timeParts.length == 3) {
              return DateTime(
                int.parse(dateParts[0]), // year
                int.parse(dateParts[1]), // month
                int.parse(dateParts[2]), // day
                int.parse(timeParts[0]), // hour
                int.parse(timeParts[1]), // minute
                int.parse(timeParts[2]), // second
              );
            }
          }
        } catch (e) {
          print('[ItemApiService] Error parsing datetime: $e');
        }
      }
    }

    // Fallback to 7 days from now
    return DateTime.now().add(PricingConfig.listingDuration);
  }

  /// Check if the latest item's payment has been completed
  ///
  /// Returns true if the item's sale_status == 1 (sold),
  /// false if sale_status == 0 (not sold),
  /// or null if an error occurs.
  ///
  /// This is used for polling to detect when a payment completes.
  static Future<bool?> checkPaymentStatus() async {
    try {
      final response = await http.get(
        Uri.parse(ApiConfig.getItemUrl),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode != 200) {
        print(
          '[ItemApiService] Payment status check failed: ${response.statusCode}',
        );
        return null;
      }

      final jsonData = jsonDecode(response.body);

      // Backend returns { "status": true } when item is sold
      // and { "status": false } when item is not sold
      return jsonData['status'] == true;
    } catch (e) {
      print('[ItemApiService] Error checking payment status: $e');
      return null;
    }
  }

  /// Update the price of the latest item
  ///
  /// Creates a new Stripe price and payment link for the item.
  /// Returns the updated Item with the new payment link, or null if an error occurs.
  ///
  /// This is used when the dynamic pricing changes significantly and needs
  /// to be reflected in the backend and payment system.
  static Future<Item?> updateItemPrice(double newPrice) async {
    try {
      print(
        '[ItemApiService] Updating item price to \$${newPrice.toStringAsFixed(2)}',
      );

      final response = await http.put(
        Uri.parse('${ApiConfig.baseUrl}/api/item/price'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'price': newPrice}),
      );

      if (response.statusCode != 200) {
        throw Exception(
          'Failed to update price: ${response.statusCode} ${response.body}',
        );
      }

      final jsonData = jsonDecode(response.body);

      if (jsonData['items'] == null || jsonData['items'].isEmpty) {
        print('[ItemApiService] No item data in update response');
        return null;
      }

      // Parse the updated item
      final updatedItemData = jsonData['items'][0];
      return _parseItemFromBackend(updatedItemData, isSold: false);
    } catch (e) {
      print('[ItemApiService] Error updating price: $e');
      return null;
    }
  }

  /// Call the return endpoint to cancel a test purchase
  /// Returns true if return was successful, false otherwise
  static Future<bool> returnItem() async {
    try {
      print('[ItemApiService] Calling return endpoint');

      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/api/return'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        print('[ItemApiService] Item returned successfully');
        return true;
      } else {
        print(
          '[ItemApiService] Return failed: ${response.statusCode} ${response.body}',
        );
        return false;
      }
    } catch (e) {
      print('[ItemApiService] Error returning item: $e');
      return false;
    }
  }
}

File: bumpbox/src/services/itemApiService.ts
import axios from "axios";
import { API_ENDPOINTS, apiUrl } from "../api/config";
import { LISTING_DURATION_MS } from "../config/pricingConfig";
import type { Item } from "../types/item";
import {
    ItemFetchStatus,
    type ItemFetchResult,
} from "../types/item";

function parsePrice(priceData: unknown): number {
    if (priceData == null) return 100.0;
    if (typeof priceData === "number") return priceData;
    if (typeof priceData === "string") {
        const parsed = parseFloat(priceData);
        return Number.isNaN(parsed) ? 100.0 : parsed;
    }
    return 100.0;
}

function parseDateTime(dateData: unknown): Date {
    if (dateData == null) {
        return new Date(Date.now() + LISTING_DURATION_MS);
    }

    if (typeof dateData === "string") {
        const isoParsed = new Date(dateData);
        if (!Number.isNaN(isoParsed.getTime())) return isoParsed;

        const parts = dateData.split(" ");
        if (parts.length === 2) {
            const dateParts = parts[0].split("-");
            const timeParts = parts[1].split(":");
            if (dateParts.length === 3 && timeParts.length === 3) {
                return new Date(
                    parseInt(dateParts[0], 10),
                    parseInt(dateParts[1], 10) - 1,
                    parseInt(dateParts[2], 10),
                    parseInt(timeParts[0], 10),
                    parseInt(timeParts[1], 10),
                    parseInt(timeParts[2], 10),
                );
            }
        }
    }

    return new Date(Date.now() + LISTING_DURATION_MS);
}

function parseItemFromBackend(data: Record<string, unknown>, isSold = false): Item {
    const itemId = data.itemid?.toString() ?? "unknown";
    const itemName = (data.item_name as string) ?? "Unknown Item";
    const price = parsePrice(data.price);
    const description =
        (data.description as string) ??
        "High-quality item available for purchase";
    const paymentLink = (data.paymentLink as string) ?? null;
    const expirationDate = parseDateTime(data.datetime_expire);
    const listedAt = new Date(expirationDate.getTime() - LISTING_DURATION_MS);

    return {
        id: itemId,
        name: itemName,
        description,
        startingPrice: price * 1.15,
        floorPrice: price,
        listedAt,
        listingDurationMs: LISTING_DURATION_MS,
        paymentLink,
        isSold,
    };
}

export async function fetchLatestItem(): Promise<ItemFetchResult> {
    try {
        const response = await axios.get(apiUrl(API_ENDPOINTS.getItem), {
            headers: { "Content-Type": "application/json" },
            validateStatus: (status) => status === 200 || status === 404,
        });

        if (response.status === 404) {
            return { status: ItemFetchStatus.Error, message: "No item found" };
        }

        const jsonData = response.data as Record<string, unknown>;
        const status = jsonData.status;
        const message = jsonData.message?.toString() ?? "";

        if (status === true && message === "Item is sold with status sold") {
            return { status: ItemFetchStatus.EmptyLocker, message };
        }

        if (jsonData.data == null) {
            return { status: ItemFetchStatus.Error, message: "No item data" };
        }

        const isSold = status === true && message === "Item is sold with status sold";
        const item = parseItemFromBackend(
            jsonData.data as Record<string, unknown>,
            isSold,
        );

        return { status: ItemFetchStatus.ItemAvailable, item };
    } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return { status: ItemFetchStatus.Error, message };
    }
}

export async function checkPaymentStatus(): Promise<boolean | null> {
    try {
        const response = await axios.get(apiUrl(API_ENDPOINTS.getItem), {
            headers: { "Content-Type": "application/json" },
        });
        const jsonData = response.data as Record<string, unknown>;
        return jsonData.status === true;
    } catch {
        return null;
    }
}

export async function updateItemPrice(newPrice: number): Promise<Item | null> {
    try {
        const response = await axios.put(
            apiUrl(API_ENDPOINTS.updatePrice),
            { price: newPrice },
            { headers: { "Content-Type": "application/json" } },
        );

        const jsonData = response.data as Record<string, unknown>;
        const items = jsonData.items as Record<string, unknown>[] | undefined;

        if (!items?.length) return null;

        return parseItemFromBackend(items[0], false);
    } catch {
        return null;
    }
}

export async function returnItem(): Promise<boolean> {
    try {
        const response = await axios.get(apiUrl(API_ENDPOINTS.returnItem), {
            headers: { "Content-Type": "application/json" },
        });
        return response.status === 200;
    } catch {
        return false;
    }
}

export async function triggerBuyWebhook(testing: boolean): Promise<boolean | null> {
    try {
        const response = await axios.post(
            apiUrl(API_ENDPOINTS.webhook),
            { testing_intent: testing },
            { headers: { "Content-Type": "application/json" } },
        );
        return response.status === 200 ? true : null;
    } catch {
        return null;
    }
}

export async function getSolenoidState(): Promise<boolean | null> {
    try {
        const response = await axios.get(apiUrl(API_ENDPOINTS.solenoidState), {
            headers: { "Content-Type": "application/json" },
        });
        const data = response.data as Record<string, unknown>;
        return data.solenoidOn === true;
    } catch {
        return null;
    }
}

export async function toggleSolenoid(): Promise<boolean | null> {
    try {
        const response = await axios.post(
            apiUrl(API_ENDPOINTS.solenoidToggle),
            {},
            { headers: { "Content-Type": "application/json" } },
        );
        const data = response.data as Record<string, unknown>;
        return data.solenoidOn === true;
    } catch {
        return null;
    }
}

export function isValidPhoneNumber(phone: string): boolean {
    const cleaned = phone.replace(/[\s-]/g, "");

    if (cleaned.startsWith("+65")) {
        return cleaned.length === 11 && /^\+65[0-9]{8}$/.test(cleaned);
    }
    if (cleaned.startsWith("65")) {
        return cleaned.length === 10 && /^65[0-9]{8}$/.test(cleaned);
    }
    return cleaned.length === 8 && /^[0-9]{8}$/.test(cleaned);
}

export function formatPhoneNumber(phone: string): string {
    const cleaned = phone.replace(/[\s-]/g, "");

    if (cleaned.startsWith("+65")) return cleaned;
    if (cleaned.startsWith("65")) return `+${cleaned}`;
    return `+65${cleaned}`;
}

export async function createItemListing(params: {
    phone: string;
    itemName: string;
    description: string;
    price: number;
    days: number;
}): Promise<{ itemId: string; paymentLink: string }> {
    const response = await axios.post(
        apiUrl(API_ENDPOINTS.createItem),
        {
            phone: params.phone,
            item_name: params.itemName,
            description: params.description,
            price: params.price,
            days: params.days,
        },
        { headers: { "Content-Type": "application/json" } },
    );

    if (response.status !== 201) {
        throw new Error(`Failed to create item: ${response.status}`);
    }

    const data = response.data as Record<string, unknown>;
    if (data.message !== "Item created successfully") {
        throw new Error(`Unexpected response: ${data.message}`);
    }

    const itemData = data.data as Record<string, unknown> | undefined;

    return {
        itemId: String(data.itemId ?? ""),
        paymentLink: (itemData?.paymentLink as string) ?? "",
    };
}
